### PR TITLE
Add Flax training example

### PR DESCRIPTION
**Pull request by awolant**
## Category:

**New feature** 

## Description:

New tutorial showing how to train neural network implemented in Flax with DALI. It builds on the JAX training tutorial - significant parts are the same. 

### Affected modules and functionalities:

JAX documentation.

### Key points relevant for the review:

Is it understandable? Spelling, grammar?

### Tests:

- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [x] RST
  - [x] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-3564

### CODE REVIEW

**Comment by awolant** [comment]
!build

File: docs/examples/frameworks/jax/flax-basic_example.ipynb [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Training neural network with DALI and Flax\n",
+    "\n",
+    "This simple example shows how to train a neural network implemented in Flax with DALI pipelines. If you want to learn more about training neural networks with Flax look into [Flax Getting Started](https://flax.readthedocs.io/en/latest/getting_started.html) example.\n",
+    "\n",
+    "DALI setup is very similar to the [training example with pure JAX](jax-basic_example.ipynb). The only difference is the addition of a trailing dimention to the returned image to make it compatible with Flax convolutions. If you are familiar with how to use DALI with JAX you can skip this part and move to the training section of this notebook.\n",
```
Comment (banasraf) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:12]:
```suggestion
    "DALI setup is very similar to the [training example with pure JAX](jax-basic_example.ipynb). The only difference is the addition of a trailing dimension to the returned image to make it compatible with Flax convolutions. If you are familiar with how to use DALI with JAX, you can skip this part and move to the training section of this notebook.\n",
```

File: docs/examples/frameworks/jax/flax-basic_example.ipynb [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+    "\n",
+    "**Here is a quick explnation of how these parameters work:**\n",
+    "\n",
+    " - `output_map`: iterators return a dictionary with outputs of the pipeline as its values. Keys in this dictionary are defined by `output_map`. For example, `labels` output returned from the DALI pipeline defined above will be accessible as `iterator_output['labels']`,\n",
+    " - `reader_name`: setting this parameter introduces the notion of an epoch to our iterator. DALI pipeline itself is infinite, it will return the data indefinately, wrapping around the dataset. DALI readers (such as `fn.readers.caffe2` used in this example) have access to the information about the size of the dataset. If we want to pass this information to the iterator we need to point to the operator that should be queried for the dataset size. We do it by naming the operator (note `name=\"mnist_caffe2_reader\"`) and passing the same name as the value for `reader_name` argument,\n",
```
Comment (banasraf) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:143]:
```suggestion
    " - `reader_name`: setting this parameter introduces the notion of an epoch to our iterator. DALI pipeline itself is infinite, it will return the data indefinately, wrapping around the dataset. DALI readers (such as `fn.readers.caffe2` used in this example) have access to the information about the size of the dataset. If we want to pass this information to the iterator, we need to point to the operator that should be queried for the dataset size. We do it by naming the operator (note `name=\"mnist_caffe2_reader\"`) and passing the same name as the value for `reader_name` argument,\n",
```

File: docs/examples/frameworks/jax/flax-basic_example.ipynb [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With the setup above DALI iterators are ready for the training. \n",
+    "\n",
+    "Now we need to setup model and training utilities. The goal of this notebook is not to explain Flax concepts. We want to show how to train models implemented in Flax with DALI as a data loading and preprocessing library. We used standard Flax tools do define simple neural network. We have functions to create an instance of this network, run one training step on it and calculate accuracy of the model at the end of each epoch.\n",
+    "\n",
+    "If you want to learn more about Flax and get better understanding of the code below look into [Flax Documentation](https://flax.readthedocs.io/en/latest/)."
```
Comment (banasraf) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:201]:
```suggestion
    "If you want to learn more about Flax and get better understanding of the code below, look into [Flax Documentation](https://flax.readthedocs.io/en/latest/)."
```

File: docs/examples/frameworks/jax/flax-basic_example.ipynb [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    "momentum = 0.9\n",
+    "\n",
+    "model_state = create_model_state(init_rng, learning_rate, momentum)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "At this point everything is ready to run the training."
```
Comment (banasraf) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:306]:
```suggestion
    "At this point, everything is ready to run the training."
```

File: docs/examples/frameworks/jax/flax-basic_example.ipynb [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    "print(f\"Number of batches in training iterator = {len(training_iterator)}\")\n",
+    "print(f\"Number of batches in validation iterator = {len(validation_iterator)}\")"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With the setup above DALI iterators are ready for the training. \n",
```
Comment (banasraf) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:197]:
```suggestion
    "With the setup above, DALI iterators are ready for the training. \n",
```

File: docs/examples/frameworks/jax/flax-basic_example.ipynb [hunk 6]
```diff
@@ -0,0 +1,368 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Training neural network with DALI and Flax\n",
+    "\n",
+    "This simple example shows how to train a neural network implemented in Flax with DALI pipelines. If you want to learn more about training neural networks with Flax look into [Flax Getting Started](https://flax.readthedocs.io/en/latest/getting_started.html) example.\n",
```
Comment (banasraf) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:10]:
```suggestion
    "This simple example shows how to train a neural network implemented in Flax with DALI pipelines. If you want to learn more about training neural networks with Flax, look into [Flax Getting Started](https://flax.readthedocs.io/en/latest/getting_started.html) example.\n",
```

File: docs/examples/frameworks/jax/flax-basic_example.ipynb [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+    "\n",
+    "**Here is a quick explnation of how these parameters work:**\n",
+    "\n",
+    " - `output_map`: iterators return a dictionary with outputs of the pipeline as its values. Keys in this dictionary are defined by `output_map`. For example, `labels` output returned from the DALI pipeline defined above will be accessible as `iterator_output['labels']`,\n",
+    " - `reader_name`: setting this parameter introduces the notion of an epoch to our iterator. DALI pipeline itself is infinite, it will return the data indefinately, wrapping around the dataset. DALI readers (such as `fn.readers.caffe2` used in this example) have access to the information about the size of the dataset. If we want to pass this information to the iterator we need to point to the operator that should be queried for the dataset size. We do it by naming the operator (note `name=\"mnist_caffe2_reader\"`) and passing the same name as the value for `reader_name` argument,\n",
+    "  - `auto_reset`: this argument controls the behaviour of the iterator after the end of an epoch. If set to `True` will automatically reset the state of the iterator and prepare it to start the next epoch."
```
Comment (banasraf) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:144]:
```suggestion
    "  - `auto_reset`: this argument controls the behaviour of the iterator after the end of an epoch. If set to `True`, it will automatically reset the state of the iterator and prepare it to start the next epoch."
```

File: docs/examples/frameworks/jax/flax-basic_example.ipynb [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+    "\n",
+    "    return correct_predictions / iterator.size"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With utilities defined above we can create an instance of the model we want to train."
```
Comment (banasraf) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:277]:
```suggestion
    "With utilities defined above, we can create an instance of the model we want to train."
```

**Review by banasraf — APPROVED** [review]

File: docs/examples/frameworks/jax/flax-basic_example.ipynb (hunk 1 above)
Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb]: Done

File: docs/examples/frameworks/jax/flax-basic_example.ipynb (hunk 2 above)
Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb]: Done

File: docs/examples/frameworks/jax/flax-basic_example.ipynb (hunk 3 above)
Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb]: Done

File: docs/examples/frameworks/jax/flax-basic_example.ipynb (hunk 4 above)
Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb]: Done

File: docs/examples/frameworks/jax/flax-basic_example.ipynb (hunk 8 above)
Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb]: Done

File: docs/examples/frameworks/jax/flax-basic_example.ipynb (hunk 5 above)
Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb]: Done

File: docs/examples/frameworks/jax/flax-basic_example.ipynb (hunk 7 above)
Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb]: Done

**Comment by awolant** [comment]
Some of the comments were also applicable to the basic JAX tutorial since it shares some content with this one. I applied them there as well.

File: docs/examples/frameworks/jax/flax-basic_example.ipynb [hunk 9]
```diff
@@ -0,0 +1,368 @@
+{
```
Comment (szalpal) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:1]: <p>When I clicked the "training example with pure JAX" link, it redirected me to raw output of the jupyter notebook. Is this intentional? Or maybe the reviewing tool fails here?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/4978/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (szalpal) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:1]: <p>In this section "Getting started" link redirected me to the raw jupyter notebook and "pipeline documentation" to the raw rst file. Could you double-check these?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/4978/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (szalpal) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:1]: <p>Maybe you could also explain above, why setting <code>seed=0</code>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/4978/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (szalpal) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:1]: <p>Suggestion:</p><p>"Next step is to instantiate pipelines and build them." -&gt; "Next step is to instantiate DALI pipelines and build them." </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/4978/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (szalpal) [inline docs/examples/frameworks/jax/flax-basic_example.ipynb:1]: <p>Suggestions:</p><p>"For DALI pipeline to work with JAX it needs to be wrapped with appropriate DALI iterator." -&gt; "For DALI pipeline to work with JAX, the former needs to be wrapped with an appropriate DALI iterator."</p><p> </p><p>"In addition to the pipeline we can pass the" -&gt; "In addition to the DALI pipeline object we can pass the"</p><p> </p><p>"<strong>Here is a quick explnation of how these parameters work:" </strong>- maybe it would be good to put a link to the full documentation somewhere here?</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/4978/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

**Review by szalpal — APPROVED** [review]
Please double-check the links, cause they did not work correctly in the ReviewNB (but it might be a problem of the review tool)

File: docs/examples/frameworks/jax/flax-basic_example.ipynb (hunk 9 above)
Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb:1]: Yes, the review tool can't handle these links. I checked them in the docs build.

Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb:1]: Yes, the review tool can't handle these links. I checked them in the docs build.

Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb:1]: Done

Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb:1]: Done

Comment (awolant) [reply docs/examples/frameworks/jax/flax-basic_example.ipynb:1]: This is described in the tutorials that are linked just above so I am going to leave it to keep the notebook as lean as possible and only related to JAX+DALI.

**Comment by awolant** [comment]
!build
